File: src/main/java/net/atos/zac/zoeken/model/ZoekParameters.java
/*
 * SPDX-FileCopyrightText: 2022 Atos
 * SPDX-License-Identifier: EUPL-1.2+
 */

package net.atos.zac.zoeken.model;

import java.util.ArrayList;
import java.util.EnumMap;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Set;

import net.atos.zac.shared.model.SorteerRichting;
import net.atos.zac.zoeken.model.index.ZoekObjectType;

public class ZoekParameters {

    private int rows;

    private int start;

    private final ZoekObjectType type;

    private EnumMap<ZoekVeld, String> zoeken = new EnumMap<>(ZoekVeld.class);

    private EnumMap<DatumVeld, DatumRange> datums = new EnumMap<>(DatumVeld.class);

    private final EnumMap<FilterVeld, FilterParameters> filters = new EnumMap<>(FilterVeld.class);

    private final HashMap<String, String> filterQueries = new HashMap<>();

    private Sortering sortering = new Sortering(SorteerVeld.CREATED, SorteerRichting.DESCENDING);

    public ZoekParameters(final ZoekObjectType type) {
        this.type = type;
        this.getBeschikbareFilters().forEach(filterVeld -> this.addFilter(filterVeld, new FilterParameters(new ArrayList<>(), false)));
    }

    public int getRows() {
        return rows;
    }

    public void setRows(final int rows) {
        this.rows = rows;
    }

    public int getStart() {
        return start;
    }

    public void setStart(final int start) {
        this.start = start;
    }

    public Map<ZoekVeld, String> getZoeken() {
        return zoeken;
    }

    public void setZoeken(final EnumMap<ZoekVeld, String> zoeken) {
        this.zoeken = zoeken;
    }

    public void addZoekVeld(final ZoekVeld zoekVeld, final String zoekTekst) {
        this.zoeken.put(zoekVeld, zoekTekst);
    }

    public EnumMap<DatumVeld, DatumRange> getDatums() {
        return datums;
    }

    public void setDatums(final EnumMap<DatumVeld, DatumRange> datums) {
        this.datums = datums;
    }

    public void addDatum(final DatumVeld zoekVeld, final DatumRange range) {
        this.datums.put(zoekVeld, range);
    }


    public EnumMap<FilterVeld, FilterParameters> getFilters() {
        return filters;
    }


    public void addFilter(FilterVeld veld, String waarde) {
        this.filters.put(veld, new FilterParameters(List.of(waarde), false));
    }

    public void addFilter(FilterVeld veld, FilterParameters filterParameters) {
        this.filters.put(veld, filterParameters);
    }

    public void addFilterQuery(final String veld, final String waarde) {
        this.filterQueries.put(veld, waarde);
    }

    public HashMap<String, String> getFilterQueries() {
        return filterQueries;
    }

    public Sortering getSortering() {
        return sortering;
    }

    public void setSortering(final SorteerVeld veld, SorteerRichting richting) {
        this.sortering = new Sortering(veld, richting);
    }

    public ZoekObjectType getType() {
        return type;
    }

    public boolean isGlobaalZoeken() {
        return this.type == null;
    }

    private Set<FilterVeld> getBeschikbareFilters() {
        if (type == null) {
            return FilterVeld.getFacetten();
        }
        return switch (type) {
            case ZAAK -> FilterVeld.getZaakFacetten();
            case TAAK -> FilterVeld.getTaakFacetten();
            case DOCUMENT -> FilterVeld.getDocumentFacetten();
        };
    }
}

File: src/main/app/src/app/zoeken/model/zoek-parameters.ts
/*
 * SPDX-FileCopyrightText: 2022 Atos
 * SPDX-License-Identifier: EUPL-1.2+
 */

import { DatumVeld } from "./datum-veld";
import { ZoekVeld } from "./zoek-veld";
import { FilterVeld } from "./filter-veld";
import { SorteerVeld } from "./sorteer-veld";
import { ZoekObjectType } from "./zoek-object-type";
import { DatumRange } from "./datum-range";
import { FilterParameters } from "./filter-parameters";
import { ZoekFilters } from "../../gebruikersvoorkeuren/zoekopdracht/zoekopdracht.component";

export class ZoekParameters implements ZoekFilters {
  readonly filtersType = "ZoekParameters";
  type: ZoekObjectType;
  alleenMijnZaken = false;
  alleenOpenstaandeZaken = false;
  alleenAfgeslotenZaken = false;
  alleenMijnTaken = false;
  zoeken: Partial<Record<ZoekVeld, string>> = {};
  filters: Partial<Record<FilterVeld, FilterParameters>> = {};
  datums: Partial<Record<DatumVeld, DatumRange>> = {};
  sorteerVeld: SorteerVeld;
  sorteerRichting: "desc" | "asc" | "";
  rows = 25;
  page = 0;

  static heeftActieveFilters(zoekFilters: any): boolean {
    if (zoekFilters.zoeken) {
      for (const field in zoekFilters.zoeken) {
        if (zoekFilters.zoeken.hasOwnProperty(field)) {
          if (zoekFilters.zoeken[field] != null) {
            return true;
          }
        }
      }
    }
    if (zoekFilters.filters) {
      for (const field in zoekFilters.filters) {
        if (zoekFilters.filters.hasOwnProperty(field)) {
          if (0 < zoekFilters.filters[field]?.waarden?.length) {
            return true;
          }
        }
      }
    }
    if (zoekFilters.datums) {
      for (const field in zoekFilters.datums) {
        if (zoekFilters.datums.hasOwnProperty(field)) {
          const datum: DatumRange = zoekFilters.datums[field];
          if (datum?.van != null || datum?.tot != null) {
            return true;
          }
        }
      }
    }
    return false;
  }
}
